### PR TITLE
New kernels

### PR DESCRIPTION
**Pull request by jorge-solano-avila**

### CODE REVIEW

**Comment by 0x33467** [comment]
LGTM at least in the implementation.

**Comment by targos** [comment]
Hello. Thanks a lot for the contribution!
I will take a close look to the implementation later but before, I have a general remark. We try to have a consistent coding style in all the mljs repositories and we would like to keep it here. Can I ask you please to adapt to our style ? That would mean reverting   5bbee89 and follow the original style in the other commits.
I added a command to check the style with eslint (`npm run eslint`) and another one to try to fix most of the issues automatically (`npm run eslint-fix`). The eslint command is also executed as part of the `npm test` procedure.

**Comment by maasencioh** [comment]
Hello @targos, I think that now it's good

File: src/kernel.js [hunk 1]
```diff
@@ -4,26 +4,36 @@ const Matrix = require('ml-matrix');
...
+const CauchyKernel = require('./kernels/cauchy-kernel');
+const ExponentialKernel = require('./kernels/exponential-kernel');
+const HistogramKernel = require('./kernels/histogram-intersection-kernel');
+const LaplacianKernel = require('./kernels/laplacian-kernel');
+const MultiquadraticKernel = require('./kernels/multiquadratic-kernel');
+const RationalKernel = require('./kernels/rational-quadratic-kernel');
 
 class Kernel {
     constructor(type, options) {
+        var kernelType = {  'gaussian': new GaussianKernel(options),
```
Comment (targos) [inline src/kernel.js]: This is not very efficient. You are creating a lot of instances just to use one of them in the end. I suggest you do it like that instead:

``` js
// outside of the Kernel constructor
const kernelTypes = {
    gaussian: GaussianKernel,
    rbf: GaussianKernel,
    polynomial: PolynomialKernel,
    // ...
};

class Kernel {
    constructor(type, options) {
        if (typeof type === 'string') {
            type = type.toLowerCase();
            var KernelConstructor = kernelTypes[type];
            if (KernelConstructor) {
                this.kernelFunction = new KernelConstructor(options);
            } else {
                throw new Error('unsupported kernel type: ' + type);
            }
        } else if ...
    }
}
```

File: src/kernel.js [hunk 2]
```diff
@@ -4,26 +4,36 @@ const Matrix = require('ml-matrix');
...
-                    this.kernelFunction = new GaussianKernel(options);
-                    break;
-                case 'polynomial':
-                case 'poly':
-                    this.kernelFunction = new PolynomialKernel(options);
-                    break;
-                default:
-                    throw new Error('unsupported kernel type: ' + type);
-            }
-        } else if (typeof type === 'object' && typeof type.compute === 'function') {
```
Comment (targos) [inline src/kernel.js:23]: please don't remove this case and the next one. The idea is to support any kernel that the user may provide

File: test/kernel.js [hunk 3]
```diff
@@ -1,24 +1,65 @@
-'use strict';
+"use strict";
```
Comment (targos) [inline test/kernel.js]: The style was not corrected in this file ;)
I forgot to add the test directory to the eslint command

File: src/kernels/anova-kernel.js [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+        this.sigma = options.sigma;
+        this.degree = options.degree;
+    }
+
+    compute(x, y) {
+        var sum = 0;
+        var len = Math.min(x.length, y.length);
+        for (var i = 1; i <= len; ++i) {
+            sum += Math.pow(Math.exp(-this.sigma * Math.pow(Math.pow(x[i - 1], i)
+                    - Math.pow(y[i - 1], i), 2)), this.degree);
```
Comment (targos) [inline src/kernels/anova-kernel.js]: I think it's clearer if you put the operator at the end of the previous line but you can leave it here if you disagree.

**Comment by jorge-solano-avila** [comment]
Hello @targos, thanks for the feedback. I changed the project according to your suggestions

**Comment by targos** [comment]
LGTM

**Comment by targos** [comment]
Merged. Thanks a lot !

**Pull request merged**